crates/interpreter/src/instructions/memory.rs

```diff
@@ -35,3 +37,26 @@ pub fn msize(interpreter: &mut Interpreter, _host: &mut dyn Host) {
     gas!(interpreter, gas::BASE);
     push!(interpreter, U256::from(interpreter.memory.effective_len()));
 }
+// From EIP-5656 MCOPY
+pub fn mcopy<SPEC: Spec>(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    check!(interpreter, SPEC::enabled(CANCUN));           
+    pop!(interpreter, dest, src, len);
+    let len = as_usize_or_fail!(interpreter, len, InstructionResult::InvalidOperandOOG);
+    gas_or_fail!(interpreter, gas::verylowcopy_cost(len as u64));
+    let dest = as_usize_or_fail!(
+        interpreter,
+        dest,
+        InstructionResult::InvalidOperandOOG
+    );
+    let src = as_usize_or_fail!(
+        interpreter,
+        src,
+        InstructionResult::InvalidOperandOOG
+    );
+    let (data_end, overflow) = dest.overflowing_add(len);
+    if overflow || data_end > interpreter.return_data_buffer.len() {
+        interpreter.instruction_result = InstructionResult::OutOfOffset;
+        return;
+    }
+    interpreter.memory.set_data(src, dest, len, &interpreter.contract.input);
```

Comment:
And here we are coping the `contract.input` that is different from the thing we check above

Reply:
Yep this is right, i read the memory from slice to pass the right data here